internal/primitives/state-machine/changeset.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2025 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package statemachine
+
+import (
+	"iter"
+
+	"github.com/tidwall/btree"
+)
+
+// From btree.Set constraints
+type ordered interface {
+	~int | ~int8 | ~int16 | ~int32 | ~int64 |
+		~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr |
+		~float32 | ~float64 | ~string
+}
+
+// Describes in which mode the node is currently executing.
+type ExecutionMode = uint8
+
+const (
+	// Executing in client mode: Removal of all transactions possible.
+	ExecutionModeClient = iota
+	// Executing in runtime mode: Transactions started by the client are protected.
+	ExecutionModeRuntime
+)
+
+// Dirty keys are a set of keys that have been modified in each transaction.
+type DirtyKeysSets[K ordered] []btree.Set[K]
```

Comment:
Pretty sure multiple keys are implied with the term "keyset".  Do we need to export out of `statemachine`?
```suggestion
type dirtyKeySets[K ordered] []btree.Set[K]
```